p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,398 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-msgio/protoio"
+	"github.com/pion/datachannel"
+	"github.com/pion/webrtc/v3"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+)
+
+var _ network.MuxedStream = &dataChannel{}
+
+const (
+	// maxMessageSize is limited to 16384 bytes in the SDP.
+	maxMessageSize uint64 = 16384
+	// Max message size limit in the SDP is limited to 16384 bytes.
+	// We keep a maximum of 2 messages in the buffer
+	maxBufferedAmount uint64 = 3 * maxMessageSize
+	// bufferedAmountLowThreshold and maxBufferedAmount are bound
+	// to a stream but congestion control is done on the whole
+	// SCTP association. This means that a single stream can monopolize
+	// the complete congestion control window (cwnd) if it does not
+	// read stream data and it's remote continues to send. We can
+	// add messages to the send buffer once there is space for 1 full
+	// sized message.
+	bufferedAmountLowThreshold uint64 = 16384
+
+	protoOverhead  int = 5
+	varintOverhead int = 2
+)
+
+// Package pion detached data channel into a net.Conn
+// and then a network.MuxedStream
+type dataChannel struct {
+	channel       *webrtc.DataChannel
+	rwc           datachannel.ReadWriteCloser
+	laddr         net.Addr
+	raddr         net.Addr
+	readDeadline  time.Time
+	writeDeadline time.Time
+
+	closeWriteOnce sync.Once
+	closeReadOnce  sync.Once
+	resetOnce      sync.Once
+
+	state channelState
+
+	ctx            context.Context
+	cancel         context.CancelFunc
+	m              sync.Mutex
+	readBuf        []byte
+	writeAvailable chan struct{}
+	reader         protoio.Reader
+	writer         protoio.Writer
+
+	requestRead     chan struct{}
+	receivedMessage chan struct{}
+	deadlineUpdated chan struct{}
+
+	wg sync.WaitGroup
+}
+
+func newDataChannel(
+	channel *webrtc.DataChannel,
+	rwc datachannel.ReadWriteCloser,
+	pc *webrtc.PeerConnection,
+	laddr, raddr net.Addr) *dataChannel {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	result := &dataChannel{
+		channel:         channel,
+		rwc:             rwc,
+		laddr:           laddr,
+		raddr:           raddr,
+		readDeadline:    time.Time{},
+		writeDeadline:   time.Time{},
+		ctx:             ctx,
+		cancel:          cancel,
+		writeAvailable:  make(chan struct{}),
+		reader:          protoio.NewDelimitedReader(rwc, 16384),
+		writer:          protoio.NewDelimitedWriter(rwc),
+		readBuf:         []byte{},
+		requestRead:     make(chan struct{}, 5),
+		receivedMessage: make(chan struct{}, 5),
+		deadlineUpdated: make(chan struct{}),
+	}
+
+	channel.SetBufferedAmountLowThreshold(bufferedAmountLowThreshold)
+	channel.OnBufferedAmountLow(func() {
+		result.writeAvailable <- struct{}{}
+	})
+
+	result.wg.Add(1)
+	go result.readLoop()
```

Comment:
This is necessary. I put the reads in a separate goroutine so that we can continue to read in a non-blocking way in the `Write` function after the stream closes the read side. This is to continue receiving flags.  https://github.com/libp2p/go-libp2p/pull/1655/files#diff-12d09fc94970201a209887333bd4e4d408abb3c047fb4afd4f1f1b4602edec4fR166

Reply:
As a general theme in this PR, things like these need to be documented in the code. Unexpected design decisions need to be explained.

I'm not sure if this is really necessary though. While the user is still `Read`ing, we definitely don't need that Go routine, as the user is driving reads. After the user has stopped reading (either by reading the `io.EOF` or by canceling reads), we now need to be listening for a potential STOP_READING on the stream. At the very least, we should delay spawning that Go routine until that moment. This would give us a some protection against DoS attacks against this code path.

We could also just drive the reading part from calls to `Write` (after the user has stopped reading), i.e. before actually performing the write, we'd consume all messages that have accumulated on the datachannel, so we can check for the STOP_READING flag.

Reply:
The issue with the first suggestion is that we will not be able to enforce deadlines on the `Read` call. As we have discussed previously, the pion API does not currently allow for setting deadlines on `Read` or `Write`. Having the `Read` in a separate Goroutine allows the user to set a read deadline at the libp2p level. Also, the routine does not continously read from the datachannel, and waits for a signal on the `readRequested` channel.

We do drive reads from the `Write` call. The problem with directly calling `Read` on the datachannel from `Write` is that it will block until data is available.